compiler/rustc_hir_typeck/src/method/probe.rs

```diff
@@ -1093,7 +1086,7 @@ impl<'a, 'tcx> ProbeContext<'a, 'tcx> {
         &self,
         step: &CandidateStep<'tcx>,
         self_ty: Ty<'tcx>,
-        unstable_candidates: Option<&mut Vec<(Candidate<'tcx>, Symbol)>>,
+        unstable_candidates: &mut Option<Vec<(Candidate<'tcx>, Symbol)>>,
```

Comment:
Why did you invert this?

I think `Option<&mut T>` is generally preferred over `&mut Option<T>`, unless you intend to modify the value of the `Option` itself. Passing `&mut Option<T>` implies the wrong intention with this IMO, since passing in `unstable_candidates` indicates that that the caller wants unstable candidates recorded -- I don't expect the callee to ever, e.g., set `*unstable_candiates = Some(vec![])`.

Reply:
Fair enough - I'll close this PR.

It happens to simplify some of the subsequent work on arbitrary self types, but I'll find another way to do it. Most likely I'll make a newtype wrapper for this `Option`.

Reply:
I mean, I don't mind the *other* half of this PR, i.e. passing `Option<Vec<..>>` at the top level -- I don't necessarily see why these had to change from `Option<&mut Vec<>>` to `&mut Option<Vec<>>` as a consequence though?